### PR TITLE
Visual Studio: Write out DLL/EXE checksums

### PR DESCRIPTION
**Pull request by nirbheek**
commit 12cb82307f2016e19a2634a2081a9723acd27319:

```
linkers: Clarify a comment about rspfiles
```

commit 3a05548cdd381b0ad99e29050f73337d8c2fcfb3:

```
linkers: Accept both str and List[str] for _apply_prefix
Simplifies some usage.
```

commit 360e1cdd2350182b0636d2251efe50291f8806dd:

```
vs: Write checksums in PE binaries (DLLs and EXEs)
This is needed for detecting data corruption, and its absence (or
an incorrect value) is also used as a hint by anti-viruses that the
binary may be malware.

Flag is only supported by MSVC `link.exe`, not `lld-link.exe`
```
https://docs.microsoft.com/en-us/cpp/build/reference/release-set-the-checksum

commit e017147d2199a85e1c6664fb8d8f39f32a389e8b:

```
tests: Add a unit test for checksums
Adds a CI dependency on the `pefile` python module.
```

commit 2c96569374cb8f87564c4fab2560b8aee4a4d263:

```
tests/windows/16: Use pefile module instead of objdump/dumpbin
The pefile module is a CI dependency now, so we can use that instead
of objdump/dumpbin which greatly simplifies the test. Of course, this
module is also cross-platform so it will work if we add cross-win32 CI
at some point.
```

### CODE REVIEW

File: run_unittests.py [hunk 1]
```diff
@@ -4646,6 +4646,40 @@ def test_link_environment_variable_optlink(self):
...
+        cc = env.detect_c_compiler(MachineChoice.HOST)
+        cc_id = cc.get_id()
+        ld_id = cc.get_linker_id()
+        dll = glob(os.path.join(self.builddir, '*mycpplib.dll'))[0]
+        exe = os.path.join(self.builddir, 'cppprog.exe')
+        for f in (dll, exe):
+            pe = pefile.PE(f)
+            # Verify that a valid checksum was written
+            self.assertTrue(pe.verify_checksum(),
+                    msg='PE file: {!r}, compiler: {!r}, linker: {!r}'.format(f, cc_id, ld_id))
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[Flake8]__ continuation line under-indented for visual indent ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6451#issue-5054829))

<details>

| Rule |
| - |
| E128 |

</details>
<sub>posted by Sider</sub>

**Comment by nirbheek** [comment]
Ready for review and/or merging.

**Comment by nirbheek** [comment]
Ah, this will probably be affected by https://github.com/mesonbuild/meson/pull/6419, @jon-turney @dcbaker what do you think of this?

**Comment by jon-turney** [comment]
> vs: Pass /Brepro to the linker to omit the timestamp
> This makes EXEs and DLLs reproducible by omitting the timestamp and
> writing it out as -1. This flag is undocumented, see:

The equivalent `ld` flag is `--no-insert-timestamps`, but this was known to be dangerous to use (See https://sourceware.org/bugzilla/show_bug.cgi?id=16887)

> Adds a CI dependency on the `pefile` python module.

Cool.  If we have this available, `test cases/windows/16 gui app` should probably be updated to use that, rather than running `objdump` or `dumpbin`.

**Comment by nirbheek** [comment]
> The equivalent `ld` flag is `--no-insert-timestamps`, but this was known to be dangerous to use (See https://sourceware.org/bugzilla/show_bug.cgi?id=16887)

So maybe we should put it behind a flag or something?

> Cool. If we have this available, `test cases/windows/16 gui app` should probably be updated to use that, rather than running `objdump` or `dumpbin`.

I'll look into that too, thanks for the review!

**Comment by nirbheek** [comment]
I got rid of the `/Brepro` changes because it may not be a good idea to do it by default. I will open a separate PR for that discussion.

I also changed `test cases/windows/16 gui app` and now it uses the `pefile` module.

**Pull request merged**